defi/single-token-amm-dex/src/lib.rs

```diff
@@ -0,0 +1,914 @@
+                let candy_bucket: Bucket = match self.candy_vaults.get_mut(&candy_out_addr) {
+                    Some(vault) => vault.take(candy_out_nbr-(candy_out_nbr*self.fee/100)),
+                    None => { info!("Candy not in stock! ");
+                              std::process::abort()
+                    }
+                };
+
+                match self.meta_map.get_mut(&candy_out_addr.clone()) {
+                    Some((amnt_fee,_b,_c)) => *amnt_fee = *amnt_fee+candy_out_nbr*self.fee/100,
+                    None => std::process::abort()                  
+                };
+                
+                candy_bucket
+            }
+
+            // Calculate new candy price.
+            fn price_mod(
+                &mut self, 
+                amount: Decimal, 
+                address: Address, 
+                price: Decimal, 
+                flag: Decimal
+            ) -> Decimal {
+                let total_candy = self.candy_vaults.get(&address).unwrap().amount();
+
+                let candy_amnt = amount;
+                let new_price: Decimal;
+                
+                if flag == dec!(1) { 
+                    new_price = total_candy*price/(total_candy-candy_amnt);
+                } else { 
+                    new_price = total_candy*price/(total_candy+candy_amnt);
+                }
+
+                new_price
+            }
+
+            // Calculate candy output amount.
+            fn candy_sum(
+                &mut self, 
+                amnt_pri: Decimal, 
+                addr_pri: Address, 
+                addr_sec:Address, 
+                flag: Decimal
+            )-> Decimal {
+                let mut amount: Decimal = amnt_pri;
+
+                let price_in: Decimal = CandyDex::candyprice(self, addr_pri);
+                let price_out : Decimal = CandyDex::candyprice(self, addr_sec);
+
+                let tot_amnt_out = self.candy_vaults.get(&addr_sec).unwrap().amount();
+                let candy_out_amnt = amnt_pri*price_in/price_out;
+                
+                if flag == dec!(2) {
+                    amount = amnt_pri/price_in;
+                }
+
+                if flag == dec!(1) {
+                    let price_new = price_out*tot_amnt_out/(tot_amnt_out+candy_out_amnt);
+                    amount = price_new*candy_out_amnt/price_out;
+                } else {
+                    let price_new = 
+                        CandyDex::price_mod(self, amount, addr_pri, price_in, dec!(1));
+                    let (name,symbol,_price) = self.candy_map.get(&addr_pri).unwrap();
+                    let name_str = name.to_string();
+                    let symbol_str = symbol.to_string();
+                    self.candy_map.insert(addr_pri,(name_str,symbol_str,price_new));
+                    amount = amnt_pri/price_new;
+                }
+
+                amount
```

Comment:
This is a little bit confusing. I suggest doing like this:
```suggestion
                match flag {
                    dec!(1) => {
                        let price_new = price_out*tot_amnt_out/(tot_amnt_out+candy_out_amnt);
                        price_new*candy_out_amnt/price_out
                    },
                    _ => {
                        if flag == dec!(2) {
                            amount = amnt_pri/price_in;
                        }

                        let price_new = CandyDex::price_mod(self, amount, addr_pri, price_in, dec!(1));
                        let (name,symbol,_price) = self.candy_map.get(&addr_pri).unwrap();
                        let name_str = name.to_string();
                        let symbol_str = symbol.to_string();
                        self.candy_map.insert(addr_pri,(name_str,symbol_str,price_new));
                        amnt_pri/price_new
                    }
                }
```

Reply:
Yes, since the flag is always an integer, I feel like you should use something like `u8` instead of `Decimal`